[AUTO] Implement auto-plugin limited devices feature

mashoujiang opened the pull request:
### Details:
 - *Implement limited device list feature for AUTO plugin*

### Tickets:
 - *53296*

ilya-lavrenov commented on inference-engine/src/auto_plugin/auto_plugin.cpp:
```diff
@@ -24,20 +25,142 @@ namespace {
+        IE_THROW() << "Failed to load network by AUTO plugin";
+    }
+
+    return executableNetwork;
```
we need to wrap it with `AutoExecutableNetwork`..

mashoujiang replied:
It seems we could not able to wrap it with `AutoExecutableNetwork`, because the return type is `IE::ExecutableNetwork `, but `AutoExecutableNetwork` is derived from `ExecutableNetworkInternal`.

ilya-lavrenov replied:
ok, we are going to fix it here https://github.com/openvinotoolkit/openvino/pull/5645

mashoujiang commented on inference-engine/src/auto_plugin/auto_plugin.cpp:
```diff
@@ -24,20 +25,142 @@ namespace {
         return config;
     }
 
+    std::string GetNetworkPrecision(const InferenceEngine::CNNNetwork &network) {
+        for (auto&& layer : network.getInputsInfo()) {
+            auto precision = layer.second->getPrecision();
+            auto name = std::string(precision.name());
+            if (name == "I8") {
+                name = "INT8";
+            }
+            return name;
+        }
+        return {};
+    }
```
move comment from #5549:
ilya: this is not network precision. It's precision of input layer..
shoujiang: well, I remembered Maxim provided some code like:
auto nGraphFunc = network.getFunction();
for (auto &node : nGraphFunc->get_ordered_ops()) {
// auto type0 = node->input_value(0).get_element_type(); //input
auto type1 = node->input_value(1).get_element_type(); //weights
}
so is that expected? should we use typ1 or type0 as the network precision? there are lots of ops, should we use only the first one? thanks!
ilya: int8 network - if we have FakeQuantize operations in the graph
fp16 network - if we have convolution weights stored in fp16
fp32 network - if we have convolution weights stored in fp32

mashoujiang commented:
I closed #5549 , let's focus on this PR. Sorry to bring confuse. 
I notice only one comment from Ilya, if @myshevts @ilya-lavrenov have some review comments which are not submitted, please commit here. I have not changed any code yet.

ilya-lavrenov commented on inference-engine/src/auto_plugin/auto_plugin.cpp:
```diff
@@ -179,7 +330,26 @@ std::vector<AutoPlugin::DeviceInformation> AutoInferencePlugin::GetDeviceChoice(
+            std::vector<std::string> device_cap =
+                GetCore()->GetMetric(item, METRIC_KEY(OPTIMIZATION_CAPABILITIES));
+            for (auto &dc : device_cap) {
+                if (dc == METRIC_VALUE(BATCHED_BLOB)) {
```
why do we need to ignore it?

mashoujiang replied:
because when we use AUTO:CPU, but if the GPU is availble, we will get `BATCHED_BLOB` metric, which will cause `TEST_P(SetBlobOfKindTest, CompareWithRefs)` failed, so I think the `BATCHED_BLOB` is not supported for AUTO.
otherwise, maybe we should change the logic of current `OPTIMIZATION_CAPABILITIES`, should we make it equal to intersection for all devices?

ilya-lavrenov replied:
ok, I see your point. I think we can skip BATCHED_BLOB for now

ilya-lavrenov commented on inference-engine/src/auto_plugin/auto_plugin.cpp:
```diff
@@ -141,16 +277,31 @@ IE::Parameter AutoInferencePlugin::GetMetric(const std::string& name,
+            availableDevices.push_back(devices.substr(i, idelimeter - i));
+            i = idelimeter + 1;
+        }
+        availableDevices.push_back(devices.substr(i, devices.length() - i));
```
can we reuse `DeviceIDParser::getHeteroDevices` here?

mashoujiang replied:
yes, I will change it

ilya-lavrenov commented on inference-engine/src/auto_plugin/auto_plugin.cpp:
```diff
@@ -24,20 +25,142 @@ namespace {
+            if (capability != optCap.end() && capability->find(networkPrecision) != std::string::npos) {
+                return GPU[0];
             }
         }
```
I don't understand this logic. Looks like we need to return optimization capabilities per device, while currently it's used as a union set of optimization capabilities of AUTO sub-devices.

E.g. if GPU does not support int8, while CPU does, we still select GPU based on current logic.

mashoujiang replied:
the capability is `getCap("GPU")`, so it is single device's capabilities, it's not union.

ilya-lavrenov replied:
What I see, `optCap` are represented as `BIN`, `FP16`, `INT8` etc. And you are passing CPU as `substr`.

And `getCap` is trying to match these values and I expect it cannot do this. So,
```c++
capability != optCap.end()
```
should always be `false`

mashoujiang replied:
Sorry, I see, what I mean is old logic, the code has been changed, we will update.

ilya-lavrenov commented on inference-engine/tests/functional/plugin/cpu/shared_tests_instances/behavior/core_integration.cpp:
```diff
@@ -83,23 +83,23 @@ INSTANTIATE_TEST_CASE_P(
 INSTANTIATE_TEST_CASE_P(
         smoke_IEClassExecutableNetworkGetMetricTest, IEClassExecutableNetworkGetMetricTest_SUPPORTED_METRICS,
-        ::testing::Values("CPU", "MULTI:CPU", "HETERO:CPU", "AUTO"));
+        ::testing::Values("CPU", "MULTI:CPU", "HETERO:CPU", "AUTO:CPU"));
```
I have concern that now we tests only `AUTO:CPU` or `AUTO:GPU` while not a combination of them. E.g. all logic for device selection selects among a single device

mashoujiang replied:
I agree we need AUTO:CPU,GPU tests, since this may cause different behaviors, let me think and provide one test like AUTO:CPU,GPU. 
BTW, do you have any expected behavior for AUTO:CPU,GPU? I could add that to a test, too.

mashoujiang replied:
And what do you think if I add tests for AUTO:CPU,CPU in `openvino/inference-engine/tests/functional/plugin/shared/src/behavior`? since I may not add too much tests for AUTO:CPU,GPU, I will create a test file and write test in this folder.

ilya-lavrenov replied:
And we just instantiate existing tests for GPU with one more `AUTO:CPU,GPU`?

mashoujiang replied:
ok, maybe I could find a place to add one more `AUTO:CPU,GPU` case in existing tests. will try.

mashoujiang replied:
oh, just got what you mean..., that's brilliant, I just add tests on core_integration.cpp, I may check others files to support such test cases.

ilya-lavrenov commented on inference-engine/src/auto_plugin/auto_plugin.cpp:
```diff
@@ -24,20 +25,142 @@ namespace {
     _pluginName = "AUTO";
 }
 
+IE::ExecutableNetwork AutoInferencePlugin::LoadNetwork(const std::string &modelPath,
```
This is new override for LoadNetwork with filename

myshevts replied:
if the only diff with the LoadNEtworkExeImpl is filename passed to the core (as opposite to the cnnNetwork), maybe we need a better solution than code duplication- func template, instantiated for LoadNEtwork/LodaNEtworkExeImpl with cnnnetwork and string accordingly?

mashoujiang replied:
since the return type and function name and parameters are different, I think the best way is abstract the implementation. will have a try.

coneypo commented on inference-engine/src/auto_plugin/auto_plugin.cpp:
```diff
@@ -24,20 +25,142 @@ namespace {
+            IE_THROW(NotFound) << "No available device found";
+        }
+
+        std::sort(GPU.begin(), GPU.end(), [](const DeviceInformation& a, const DeviceInformation& b)->bool{return b.deviceName < a.deviceName;});
```
dGPU > iGPU, will add comment on it

coneypo commented on inference-engine/src/auto_plugin/auto_plugin.cpp:
```diff
@@ -24,20 +25,142 @@ namespace {
+            IE_THROW(NotFound) << "No available device found";
+        }
+
+        std::sort(GPU.begin(), GPU.end(), [](const DeviceInformation& a, const DeviceInformation& b)->bool{return b.deviceName < a.deviceName;});
```
Two  `SelectDevice` func there:

1. `SelectDevice(metaDevices)` for `LoadNetwork()`, will not check the network precision;
2. `SelectDevice(network, metaDevices, optCap)` for `LoadExeNetworkImpl()`, will check the network precesion;

mashoujiang commented on inference-engine/src/auto_plugin/auto_plugin.cpp:
```diff
@@ -24,20 +25,142 @@ namespace {
+        };
+
+        if (CPU.empty() && GPU.empty()) {
+            IE_THROW(NotFound) << "No available device found";
```
This CPU and GPU are two vectors which parse the limited candidate devices, they are not filtered by INT8 yet. 
And I agree we should always choose CPU in such situation, the below code should consider that.

mashoujiang commented on inference-engine/src/auto_plugin/auto_plugin.cpp:
```diff
@@ -179,7 +330,26 @@ std::vector<AutoPlugin::DeviceInformation> AutoInferencePlugin::GetDeviceChoice(
+            std::vector<std::string> device_cap =
+                GetCore()->GetMetric(item, METRIC_KEY(OPTIMIZATION_CAPABILITIES));
+            for (auto &dc : device_cap) {
+                if (dc == METRIC_VALUE(BATCHED_BLOB)) {
```
done

mashoujiang commented:
Hi @ilya-lavrenov @myshevts , I have updated based on your comments, I also add AUTO:CPU,GPU tests on GPU tests. The CI tests are all passed. Could you please help to have a look again? thank you very much!

ilya-lavrenov commented on inference-engine/src/auto_plugin/auto_plugin.cpp:
```diff
@@ -24,20 +26,66 @@ namespace {
+                        return "FP16";
+                }
             }
         }
```
I think we need to support not only convolution here, but it's derivatives like here https://github.com/openvinotoolkit/openvino/blob/master/inference-engine/src/transformations/src/transformations/rt_info/primitives_priority_attribute.cpp#L30-L40

Secondly, we have 2 loops actually:
1. One inside ngraph::op::util::has_op_with_type<
2. Second explicit for loop

Can we merge them together and use the second one as more generic?

mashoujiang replied:
done

ilya-lavrenov commented on inference-engine/src/auto_plugin/auto_plugin.cpp:
```diff
@@ -193,6 +281,89 @@ ConfigType AutoInferencePlugin::GetSupportedConfig(const ConfigType&  config,
+        }
+    }
+
+    auto networkPrecision = GetNetworkPrecision(network);
```
`networkPrecision` compute can be moved before `for` loop

mashoujiang replied:
done

ilya-lavrenov commented on inference-engine/src/auto_plugin/auto_plugin.cpp:
```diff
@@ -24,20 +26,66 @@ namespace {
+        auto nGraphFunc = network.getFunction();
+        bool isINTModel = ngraph::op::util::has_op_with_type<ngraph::op::FakeQuantize>(nGraphFunc);
+        if (isINTModel) {
+            return "INT8";
```
can we use `METRIC_VALUE(INT8)` instead of hardcoded string? the same for FP16 and FP32

mashoujiang replied:
done

myshevts commented on inference-engine/src/auto_plugin/auto_plugin.cpp:
```diff
@@ -193,6 +283,90 @@ ConfigType AutoInferencePlugin::GetSupportedConfig(const ConfigType&  config,
+        return metaDevices.at(0);
+    }
+
+    auto networkPrecision = GetNetworkPrecision(network);
```
if the network preciosn is the only diff from another variant of the SelectDevice, maybe it should be the (single unified) function argument (with a default value of precision::fp32)

mashoujiang replied:
got it, will modify this, thanks!

myshevts commented on inference-engine/src/auto_plugin/auto_plugin.cpp:
```diff
@@ -193,6 +283,90 @@ ConfigType AutoInferencePlugin::GetSupportedConfig(const ConfigType&  config,
+
+    for (auto&& item : GPU) {
+        std::vector<std::string> capability = GetCore()->GetMetric(item.deviceName, METRIC_KEY(OPTIMIZATION_CAPABILITIES));
+        auto res = std::find(capability.begin(), capability.end(), networkPrecision);
```
having no e.g. int8 in the capabilities doesn't imply the device is not supporting the int8 execution.
but yes, it can be less efficient, than with int8 (HW) support.

mashoujiang replied:
so is it ok? or add comment to code?

myshevts commented on inference-engine/src/inference_engine/ie_core.cpp:
```diff
@@ -9,23 +9,24 @@
-#include <ngraph/opsets/opset.hpp>
-#include <ngraph/ngraph.hpp>
 #include <ngraph/graph_util.hpp>
+#include <ngraph/ngraph.hpp>
```
why so many includes changes, while the code changes are just strings manipulations?

mashoujiang replied:
well, now we check not only Convolution but also its derivatives. like the above ilya's comment, the refer code is from: https://github.com/openvinotoolkit/openvino/blob/master/inference-engine/src/transformations/src/transformations/rt_info/primitives_priority_attribute.cpp#L30-L40

myshevts replied:
from whta I see, we don't check the convolutions in the ie_core.cpp ( this file) only inside Auto

mashoujiang replied:
sorry, I misunderstand, there should no change for headers, I think that is caused by automatically in Clion. I will modify and remove unnecessary change.

mashoujiang commented:
Hi Maxim and Ilya, I have addressed your comments, clean and simplify the code. well, I will check the CI this night later. 
BTW, since I found some unused code, I changed AutoExecutableNetwork() redundant code quickly. 
Thanks for review!

ilya-lavrenov commented on inference-engine/src/auto_plugin/auto_plugin.cpp:
```diff
@@ -12,32 +12,51 @@
+        IE_THROW() << "Please, work with AUTO device via InferencEngine::Core object";
+    }
+
+    return LoadNetworkImpl(modelPath, config);
```
I don't see `LoadNetworkImpl` implementation

mashoujiang replied:
The `LoadNetworkImpl` is a template function which is defined in the header file auto_plugin.hpp.

ilya-lavrenov commented on inference-engine/src/auto_plugin/auto_plugin.cpp:
```diff
@@ -12,32 +12,51 @@
+                std::dynamic_pointer_cast<ngraph::opset1::ConvolutionBackpropData>(node) ||
+                std::dynamic_pointer_cast<ngraph::op::ConvolutionIE>(node) ||
+                std::dynamic_pointer_cast<ngraph::op::DeconvolutionIE>(node)) {
+                auto layerType = node->input(0).get_element_type().get_type_name();
```
as far as I can see from [docs](https://docs.openvinotoolkit.org/latest/openvino_docs_ops_convolution_Convolution_1.html), convolution weights is a second argument.

mashoujiang replied:
> as far as I can see from [docs](https://docs.openvinotoolkit.org/latest/openvino_docs_ops_convolution_Convolution_1.html), convolution weights is a second argument.

OK, @coneypo please check doc and double confirm in source code.

mashoujiang replied:
I think you're right, just checked the doc opset1~opset6, and source code example `ngraph_api_examples`, it should be node->input(1). Thanks!

ilya-lavrenov commented on inference-engine/src/auto_plugin/auto_plugin.cpp:
```diff
@@ -193,6 +222,56 @@ ConfigType AutoInferencePlugin::GetSupportedConfig(const ConfigType&  config,
+    }
+
+    // Sort GPU by name: GPU.2 > GPU.1 > GPU.0 > GPU, so we always choose the GPU[0] as best device
+    std::sort(GPU.begin(), GPU.end(), [](const DeviceInformation& a, const DeviceInformation& b)->bool{return b.deviceName < a.deviceName;});
```
CC @vladimir-paramuzov are you OK with this?

myshevts replied:
so dGPU has a pri over iGPU, but in this ver we do not recognize when there are many dGPUs:
always try the one with highest index first 
If the GPU with highest index fails (e.g. due to no mem) the logic try to load the next in the list (whihc would be next dGPU... etc until we come the iGPU and finally the CPU), etc

This is "greedy" but simple

mashoujiang replied:
Yes, You're right

myshevts commented on inference-engine/src/auto_plugin/auto_plugin.cpp:
```diff
@@ -24,20 +25,142 @@ namespace {
         return config;
     }
 
+    std::string GetNetworkPrecision(const InferenceEngine::CNNNetwork &network) {
```
you have to check that there is **any** int8 first, and then return "INT8":
if (ngraph::op::util::has_op_with_type<ngraph::op::FakeQuantize>(function))
      return int8;
else 
//your current loop for fp16/fp32
...

In you  current logic, if the fp16 or fp32 conv will be met first, the func will return it.

mashoujiang replied:
Thanks Maxim, will fix it

ilya-lavrenov commented on inference-engine/src/auto_plugin/auto_plugin.cpp:
```diff
@@ -12,32 +12,51 @@
     _pluginName = "AUTO";
 }
 
+IE::ExecutableNetwork AutoInferencePlugin::LoadNetwork(const std::string &modelPath,
```
we have updated this method return value. Please, rebase and fix merge conflicts.

mashoujiang replied:
Yes, I just rebased

mashoujiang commented:
Hi @myshevts @ilya-lavrenov , thanks for your review, I just rebased to the latest master branch. 
Note that since there are some conflicts during rebase, so the last commit may not reflect the related modification very well. 
Compared to **ilya's approved version**:
1. I fixed the `GetNetworkPrecision` logic according to Maxim's comment
2. refactor the template function `LoadNetworkImpl`
3. and rebased to master branch, change related code.

ilya-lavrenov commented on inference-engine/src/auto_plugin/auto_plugin.cpp:
```diff
@@ -210,7 +177,28 @@ std::vector<AutoPlugin::DeviceInformation> AutoInferencePlugin::GetDeviceChoice(
+                // and AUTO select CPU but not GPU (GPU has this capability).
+                if (cap == METRIC_VALUE(BATCHED_BLOB)) {
+                    continue;
+                }
```
looks like we need to respect `KEY_AUTO_DEVICE_LIST` here as well.

ilya-lavrenov replied:
let's fix in future PR
